src/vivarium_census_prl_synth_pop/components/observers.py

```diff
@@ -494,3 +472,191 @@ def get_observation(self, event: Event) -> pd.DataFrame:
+
+        # increment income for all person/employment pairs with income > 0
+        income_this_time_step = pd.Series(
+            pop["income"].values * self.time_step / DAYS_PER_YEAR,
+            index=pd.MultiIndex.from_arrays(
+                [pop.index, pop["employer_id"]], names=["simulant_id", "employer_id"]
+            ),
+        )
+
+        income_this_time_step = income_this_time_step[income_this_time_step > 0]
+
+        self.income_to_date = self.income_to_date.add(income_this_time_step, fill_value=0.0)
+
+    def to_observe(self, event: Event) -> bool:
+        """Only observe if this is the final time step of the sim"""
+        tax_date = dt.datetime(event.time.year, 1, 1)
+        return self.clock() < tax_date <= event.time
+
+    def get_observation(self, event: Event) -> pd.DataFrame:
+        pop = self.population_view.get(
+            event.index,
+        )
+        pop["income"] = self.pipelines["income"](pop.index)
+
+        ### create dataframe of all person/employment pairs
+
+        # start with income to date, which has simulant_id and employer_id as multi-index
+        self.income_to_date.name = "income"  # HACK: it would be nice if this name stayed
+        # with the pd.Series, but it is getting lost at some point in the computation
+
+        df_w2 = self.income_to_date.reset_index()
+
+        # merge in simulant columns based on simulant id
+        for col in [
+            "first_name_id",
+            "middle_name_id",
+            "last_name_id",
+            "age",
+            "date_of_birth",
+            "sex",
+            "ssn",
+            "address_id",
+            "housing_type",
+        ]:
+            df_w2[col] = df_w2["simulant_id"].map(pop[col])
+
+        # for simulants without ssn, record a simulant_id for a random household
+        # member with an ssn, if one exists
+
+        simulants_wo_ssn = pd.Series(
+            df_w2[~df_w2["ssn"]].index, index=df_w2[~df_w2["ssn"]].index
+        )
+        household_members_w_ssn = (
+            pop[pop["ssn"]].groupby("address_id").apply(lambda df_g: list(df_g.index))
+        )
+        simulant_from_which_to_borrow_ssn = simulants_wo_ssn.map(household_members_w_ssn)
+        ssn_for_simulants_wo = simulant_from_which_to_borrow_ssn.dropna().map(
+            np.random.choice
+        )  # NOTE: this does not use common random numbers, is that ok here?
+
+        df_w2["ssn_id"] = ssn_for_simulants_wo
+        df_w2["ssn_id"] = df_w2["ssn_id"].fillna(-1).astype(int)
+
+        # merge in employer columns based on employer_id
+        emp = pop.groupby("employer_id").first()
+        for col in ["employer_address_id", "employer_name"]:
+            df_w2[col] = df_w2["employer_id"].map(emp[col])
+
+        # create lists of dependent ids and dependent address ids
+        df_dependents = pd.concat(
+            [
+                pop[pop["guardian_1"] != -1].assign(guardian_id=lambda x: x["guardian_1"]),
```

Comment:
Is this just concatting guardian 1 and guardian 2 rows? Can't this be done without a lambda expression?

Reply:
I want a new column called guardian_id that has the value of gaurdian_1 sometimes and guardian_2 others.  I can do it with an `.eval` if you like, e.g. `pop[pop["guardian_1"] != -1].eval("guardian_id = guardian_1")`

Reply:
I thought `.assign` was more in the style I've seen in this repo...